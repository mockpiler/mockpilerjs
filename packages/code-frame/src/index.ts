/**
 * Expose code-frame main class
 */
export * from './code-frame'

/**
 * Expose options
 */
export * from './options'
