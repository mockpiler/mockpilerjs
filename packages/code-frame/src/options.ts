export type CodeFrameOptions = {
  /**
   * Padding chars in LOCs
   */
  padding?: string

  /**
   * Separators for line numbers:
   *
   * <line-number><separator><line-of-code>
   */
  lineNumberSeparator?: string

  /**
   * Indicator for line numbers:
   *
   *  > 1|
   *    2|
   *    3|
   */
  lineNumberIndicator?: string

  /**
   * Indicator for code fragments:
   *
   *   wrong-fragment
   *   ^^^^^^^^^^^^^^
   */
  fragmentIndicator?: string

  /**
   * Lines above the highlighted line
   */
  maxTopLines?: number

  /**
   * Lines below the highlighted line
   */
  maxBottomLines?: number
}

export const defaultCodeFrameOptions: Required<CodeFrameOptions> = {
  padding: ' ',
  lineNumberSeparator: '|',
  lineNumberIndicator: '>',
  fragmentIndicator: '^',
  maxTopLines: 4,
  maxBottomLines: 4
}

export function setDefaultCodeFrameOptions(defaultOptions: CodeFrameOptions) {
  return Object.assign(defaultCodeFrameOptions, defaultOptions)
}
