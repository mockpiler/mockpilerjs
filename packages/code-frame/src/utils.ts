import type { CodeFrameOptions } from './options'

type CharRepeaterOptions = Omit<
  Required<CodeFrameOptions>,
  'maxTopLines' | 'maxBottomLines'
>

export class CharRepeater {
  private _checkedChars = new Set<keyof CharRepeaterOptions>()

  constructor(private _options: CharRepeaterOptions) {}

  get(key: keyof CharRepeaterOptions, count = 1) {
    const value = this._options[key]

    if (!this._checkedChars.has(key)) {
      if (value.length !== 1) {
        throw new Error(`Option '${key}' must be of length 1`)
      }

      this._checkedChars.add(key)
    }

    return value.repeat(count)
  }
}
