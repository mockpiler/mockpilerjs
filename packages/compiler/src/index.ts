/**
 * Expose compiler main functionalities
 */
export * from './compiler'

/**
 * Expose types
 */
export * from './types'

/**
 * Expose alias
 */
export { createCompiler as mock } from './compiler'

/**
 * Expose context utils
 */
export * from '@mockpiler/context'

/**
 * Expose transformers
 */
export * from '@mockpiler/transform'
