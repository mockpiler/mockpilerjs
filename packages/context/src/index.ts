/**
 * Expose context-related
 */
export * from './context'

/**
 * Expose utils
 */
export * from './utils'

/**
 * Expose context types
 */
export * from './types'
