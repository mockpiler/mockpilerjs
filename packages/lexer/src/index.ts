/**
 * Expose lexer-related things
 */
export * from './lexer'

/**
 * Expose useful constants
 */
export * from './constants'

/**
 * Expose token declarations
 */
export * from './token'
