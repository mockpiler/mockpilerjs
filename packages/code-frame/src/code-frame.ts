import { CodeFrameOptions, defaultCodeFrameOptions } from './options'
import { CharRepeater } from './utils'

export type CodeFrameLocation = {
  line: number
  startColumn: number
  endColumn: number
}

export function generateCodeFrame(
  source: string,
  location: CodeFrameLocation,
  options: CodeFrameOptions = {}
) {
  const mergedOptions = Object.assign(
    {},
    defaultCodeFrameOptions,
    options
  ) as Required<CodeFrameOptions>
  const charRepeater = new CharRepeater(mergedOptions)

  const startLineIndex = Math.max(
    location.line - 1 - mergedOptions.maxTopLines,
    0
  )
  const targetLineIndex = location.line - startLineIndex

  let lines = source.split(/\r?\n/)

  const maxLine = Math.min(
    location.line + mergedOptions.maxBottomLines,
    lines.length
  )

  // Limit by top/bottom max lines
  lines = lines.slice(startLineIndex, maxLine)

  lines.splice(
    targetLineIndex,
    0,
    charRepeater.get('padding', location.startColumn - 1) +
      charRepeater.get(
        'fragmentIndicator',
        location.endColumn - location.startColumn
      )
  )

  const lineNumberPrefixPadding = charRepeater.get(
    'padding',
    String(maxLine).length
  )

  const getLineNumber = (index: number) =>
    startLineIndex + index + (index > targetLineIndex ? 0 : 1)

  const getLineNumberIndicatorPrefix = (index: number) =>
    index === targetLineIndex - 1
      ? charRepeater.get('lineNumberIndicator') + charRepeater.get('padding')
      : charRepeater.get('padding', 2)

  return lines
    .map((line: any, index: any) => {
      const lineNumberIndicatorPrefix = getLineNumberIndicatorPrefix(index)

      const linePrefix = (
        lineNumberPrefixPadding +
        lineNumberIndicatorPrefix +
        (index === targetLineIndex ? '' : getLineNumber(index))
      ).slice(
        -(lineNumberPrefixPadding.length + lineNumberIndicatorPrefix.length)
      )

      return (
        linePrefix +
        charRepeater.get('lineNumberSeparator') +
        line.replace('\t', '\\t')
      )
    })
    .join('\n')
}
