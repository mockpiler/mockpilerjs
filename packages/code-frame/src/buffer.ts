import { Token } from '@mockpiler/lexer'

import { CodeFrameOptions } from './options'
import { getFragmentIndicator, getPadding } from './utils'

export class CodeFrameBuffer {
  linesBuffer: { [line: number]: string } = {}
  escapeCharsRegex: RegExp

  constructor(
    public minLine: number,
    public maxLine: number,
    public options: Required<CodeFrameOptions>
  ) {
    this.escapeCharsRegex = new RegExp(
      `(${Object.keys(options.__internal_escapedChars).join('|')})`,
      'g'
    )
  }

  hasLine(line: number) {
    return line in this.linesBuffer
  }

  addLine(line: number, ...values: any[]) {
    this.linesBuffer[line] = values.join('')
  }

  prependToLine(line: number, ...values: any[]) {
    this.linesBuffer[line] = values.join('') + this.linesBuffer[line]
  }

  appendToLine(line: number, ...values: any[]) {
    this.linesBuffer[line] += values.join('')
  }

  getFormattedOutput(highlightLocation: Token['location']) {
    const formattedCode = []
    const maxLineLength = String(this.maxLine).length
    const lineIndicatorSpacing = this.options.lineNumberIndicator.length + 1

    for (let line = this.minLine; line <= this.maxLine; line++) {
      const inHighlightLine = line === highlightLocation.start.line
      const lineValue = this.linesBuffer[line] ?? ''

      // Used as extra padding in the highlighted line
      let escapeCharsOffset = 0

      formattedCode.push(
        (inHighlightLine ? this.options.lineNumberIndicator : '') +
          getPadding(
            this.options.padding,
            (inHighlightLine
              ? this.options.lineNumberIndicator.length
              : lineIndicatorSpacing) +
              maxLineLength -
              String(line).length
          ) +
          line +
          this.options.lineNumberSeparator +
          lineValue.replace(this.escapeCharsRegex, escapeChar => {
            escapeCharsOffset++
            return `\\${this.options.__internal_escapedChars[escapeChar]}`
          })
      )

      if (inHighlightLine) {
        formattedCode.push(
          getPadding(
            this.options.padding,
            maxLineLength + lineIndicatorSpacing
          ) +
            this.options.lineNumberSeparator +
            getPadding(
              this.options.padding,
              highlightLocation.start.column - 1 + escapeCharsOffset
            ) +
            getFragmentIndicator(
              this.options.fragmentIndicator,
              highlightLocation.end.column - highlightLocation.start.column
            )
        )
      }
    }

    return formattedCode.join('\n')
  }
}
