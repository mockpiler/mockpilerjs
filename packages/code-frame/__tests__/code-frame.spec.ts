import {
  testCode,
  testOneDigitLineTokenLocation,
  testTwoDigitLineTokenLocation
} from './test-cases'
import { CodeFrameLocation, generateCodeFrame } from '../src'

const getCodeFrame = (location: CodeFrameLocation) =>
  `\n${generateCodeFrame(testCode, location)}\n`

describe('code-frame', () => {
  test('generate valid code-frame', () => {
    // Test one-digit line number
    expect(getCodeFrame(testOneDigitLineTokenLocation)).toMatchSnapshot()

    // Test two-digit line number
    expect(getCodeFrame(testTwoDigitLineTokenLocation)).toMatchSnapshot()
  })
})
