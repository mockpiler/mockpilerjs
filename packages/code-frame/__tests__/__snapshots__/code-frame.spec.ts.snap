// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`code-frame generate valid code-frame 1`] = `
"
   5|      anotherNestedValue
   6|    ]
   7|    someNestedObject: {
   8|      someNestedProperty
 > 9|      test-anotherNested: property
    |      ^^^^^^^^^^^^^^^^^^
  10|    }
  11|    anotherNestedArray: [
  12|      someNestedValue
  13|      anotherNestedValue
"
`;

exports[`code-frame generate valid code-frame 2`] = `
"
   6|    ]
   7|    someNestedObject: {
   8|      someNestedProperty
   9|      test-anotherNested: property
> 10|    }
    |    ^
  11|    anotherNestedArray: [
  12|      someNestedValue
  13|      anotherNestedValue
  14|    ]
"
`;
